### ISSUE TITLE
How to get info for type expression from raml 1 library?

### ISSUE THREAD

**Issue by ericfreese** [raml-org/webapi-parser#28]
Hi, I have a raml type library with some types defined as type expressions using other types in the file.

```raml
#%RAML 1.0 Library
types:
  Foo:
    type: object
    properties:
      foo: string
  Bar:
    type: object
    properties:
      bar: string
  FooBar: Foo | Bar
```

How can I get information on the `FooBar` type? I've tried `getDeclarationByName` but it throws saying `Declaration with name 'FooBar' not found`. Have also tried `findById` and `findByType` but those return an empty array and undefined (not sure how these differ from `getDeclarationByName` but was just stabbing in the dark based on the docs [here](https://raml-org.github.io/webapi-parser/js/classes/_webapi_parser_.webapibaseunit.html)).

Here's the full code I was trying on [runkit](https://npm.runkit.com/webapi-parser):

```js
const wap = require('webapi-parser').WebApiParser

const ramlStr = `
  #%RAML 1.0 Library
  types:
    Foo:
      type: object
      properties:
        foo: string
    Bar:
      type: object
      properties:
        bar: string
    FooBar: Foo | Bar
`

async function main () {
  const model = await wap.raml10.parse(ramlStr)

  const foo = model.getDeclarationByName('Foo');
  console.log("foo:", foo);
  
  const bar = model.getDeclarationByName('Bar');
  console.log("bar:", bar);
  
  // How do I get information on the `FooBar` type?
  const foobar = model.getDeclarationByName('FooBar');
  console.log("foobar:", foobar);
}

main()
```

And an unrelated question I'll just tack on here instead of opening a separate issue for: is there any way to iterate over all declarations found in the type library? I see some of the examples use `model.declares` but typescript (and the docs) tells me that that property doesn't exist on `WebApiBaseUnit`.

**Comment by postatum** [comment]
Hi.

> I've tried getDeclarationByName but it throws saying Declaration with name 'FooBar' not found.

Sorry, that's a bug.

For now what you can do is: call `model.declares`, see what Id a particular object has and then do `model.findById('theId')`.
E.g. in this case the id is (when parsed from a string):
```
      'id$1': 'http://a.ml/amf/default_document#/declarations/types/union/FooBar',
```
and the call it would be
```
> model.findById('http://a.ml/amf/default_document#/declarations/types/union/FooBar')
```
---
> And an unrelated question I'll just tack on here instead of opening a separate issue for: is there any way to iterate over all declarations found in the type library? I see some of the examples use model.declares but typescript (and the docs) tells me that that property doesn't exist on WebApiBaseUnit.

You can use `model.declares`. I'll check why typescript tells property doesn't exist.

**Comment by ericfreese** [comment]
Thanks! Type asserting it to WebApiModule makes declares available. Why is parse() specified in typescript as returning a WebApiBaseUnit instead of a WebApiModule?

**Comment by postatum** [comment]
> Why is parse() specified in typescript as returning a WebApiBaseUnit instead of a WebApiModule?

Because WebApiBaseUnit is a parent interface to all the things that can be parsed from a RAML document. 
You can discover a complete list of concrete classes (like WebApiModule) if go down a hierarchy at [WebApiBaseUnit API page](https://raml-org.github.io/webapi-parser/js/classes/_webapi_parser_.webapibaseunit). 
E.g. WebApiModule is under WebApiBaseUnit -> WebApiBaseUnitWithDeclaresModel -> WebApiModule.

**Comment by ericfreese** [comment]
Ok makes sense, thanks.

Another question I've run into I'll just tack on here in this thread.

How can I get the actual values of an enum?

I'm trying the following on [runkit](https://npm.runkit.com/webapi-parser) without much success. I can get the `id` of each enum value, but can't seem to find a way to get the actual string values in the enum.

```js
const wap = require('webapi-parser').WebApiParser

const ramlStr = `
  #%RAML 1.0 Library
  types:
    Foo:
      enum: [foo, bar]
`

const getAllProperties = (d) => d.graph().properties().map(p => [p, d.graph().getObjectByPropertyId(p)])

async function main () {
  const model = await wap.raml10.parse(ramlStr)

  const foo = model.declares[0];
  console.log(getAllProperties(foo));
  
  const possibleValues = foo.graph().getObjectByPropertyId('http://www.w3.org/ns/shacl#in');
  console.log(possibleValues.map(getAllProperties));
  console.log(possibleValues.map(p => p.id));
  
  // How do I get all of the actual scalar values?
  // I want an array `["foo", "bar"]`
}

main()
```

**Comment by ericfreese** [comment]
Ah, ok I've found a bunch of subclasses of model.domain.DomainElement, one of which is `ScalarNode` that has a `value` property that gives me what I need.

What is the "correct" way to programatically assert in typescript (or java/scala for that matter) that a value of type model.domain.DomainElement returned by getObjectByPropertyId is actually some more specific subclass like model.domain.ScalarNode? Does the list of strings given by graph().types() somehow relate? I'm noticing my node gives `[ 'http://a.ml/vocabularies/data#Scalar',
      'http://a.ml/vocabularies/data#Node',
      'http://a.ml/vocabularies/document#DomainElement' ]`

**Comment by postatum** [comment]
> What is the "correct" way to programatically assert in typescript (or java/scala for that matter) that a value of type model.domain.DomainElement returned by getObjectByPropertyId is actually some more specific subclass like model.domain.ScalarNode? 

I think language specific type assertions have to be used. E.g. here's how we do it in Scala: [the code](https://github.com/raml-org/webapi-parser/blob/eafaa659b1535d6aca0aba69b78659a5e54cfe6e/shared/src/main/scala/webapi/ClientConverters.scala#L54).

> Does the list of strings given by graph().types() somehow relate? I'm noticing my node gives `[ 'http://a.ml/vocabularies/data#Scalar', 'http://a.ml/vocabularies/data#Node', 'http://a.ml/vocabularies/document#DomainElement' ]`

Yes, but I'm afraid at this moment one would have to relate those types to DomainElement subclasses by hand.

**Comment by ericfreese** [comment]
Thanks @postatum.

I have another question. How can I get the type information for a property?

Here's another sample on [runkit](https://npm.runkit.com/webapi-parser):

```js
const wap = require('webapi-parser').WebApiParser

const ramlStr = `
  #%RAML 1.0 Library
  types:
    Foo:
      properties:
        foo: string
        bar: number
    Bar:
      properties:
        thing: Foo
`

// How do I get the type of a given property?
//
// For Foo.foo, I want this to return "http://www.w3.org/2001/XMLSchema#string"
// For Foo.bar, I want this to return "http://a.ml/vocabularies/shapes#number"
const getPropertyName = (property) => {
  // It looks like I can access it via the following js properties.
  // For Bar.thing though, this nested js property just contains the field name "thing"
  // How can I get some sort of string indicating that the field Bar.thing is of type Foo?
  // And what is the proper set of documented function calls to get this data?
  return property.n.ha.ec.MF.g.ha.ec.MF.g.g;

}

async function main () {
  const model = await wap.raml10.parse(ramlStr)

  const foo = model.declares[0];
  
  console.log(model.declares[0].properties.map(getPropertyName));
  console.log(model.declares[1].properties.map(getPropertyName));
}

main()
```

**Comment by postatum** [comment]
Hi @ericfreese 

For Foo properties:
```js
function getPropertyName (property) {
  return property.range.dataType.value()
}
```

For `Bar.thing`:
```js
const thingProp = model.declares[1].properties[0]
thingProp.range.linkTarget.name.value()
```

**Comment by allisonlucas** [comment]
> For Foo properties:
> 
> ```js
> function getPropertyName (property) {
>   return property.range.dataType.value()
> }
> ```
Hi @postatum, I'm working on this with @ericfreese. It doesn't appear as though the property dataType exists on range. Is there another way to access this value?

**Comment by postatum** [comment]
Original issue is fixed in a recently released version 0.2.0.
Please ask API questions/for help in new gh issues.

Thanks again @ericfreese :apple:
